### PR TITLE
Add docs for checkpointing

### PR DESCRIPTION
**Pull request by tduretz**

### CODE REVIEW

File: docs/src/IO.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+It is customary to employ [checkpointing](https://en.wikipedia.org/wiki/Application_checkpointing) during simulation that involve many time steps. A checkpoint file then needs to be written to disk. Such file allows for restarting a simulation from last checkpoint file written to disk.
+Moreover, checkpoint files may occupy a lost of disk space. Here is how to write essential particle information in a checkpoint file in [jld2 format](https://github.com/JuliaIO/JLD2.jl):
+
+```julia
+ jldsave("my_file.jld2"; 
+            particles     = Array( particles), 
+            phases        = Array( phases), 
+            phase_ratios  = Array( phase_ratios), 
+            particle_args = Array.( particle_args),
+ )
```
Comment (albert-de-montserrat) [inline docs/src/IO.md:14]:
```suggestion
jldsave(
    "my_file.jld2"; 
    particles     = Array( particles), 
    phases        = Array( phases), 
    phase_ratios  = Array( phase_ratios), 
    particle_args = Array.( particle_args),
)
```

File: docs/src/IO.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+If file size are huge, on may cast all the fields from particle structures into `Float32`. While this will spare disk space, it may hinder the reproducibility at restart. 
+
+```julia
+ jldsave("my_file.jld2"; 
+            particles     = Array( Float32, particles), 
+            phases        = Array( Float32, phases), 
+            phase_ratios  = Array( Float32, phase_ratios), 
+            particle_args = Array.( Float32, particle_args),
+ )
```
Comment (albert-de-montserrat) [inline docs/src/IO.md:26]:
```suggestion
jldsave(
    "my_file.jld2"; 
    particles     = Array( Float32, particles), 
    phases        = Array( Float32, phases), 
    phase_ratios  = Array( Float32, phase_ratios), 
    particle_args = Array.( Float32, particle_args),
)
```

File: docs/src/IO.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+## Loading a checkpoint file
+
+In order to restart a simulation, one needs to load the checkpoint file of interest. This is how to read the particle information from the checkpoint file `my_file.jld2`:
+
+```julia
+    data          = load("my_file.jld2")
+    particles     = TA(backend)(Float64, data["particles"])
+    phases        = TA(backend)(Float64, data["phases"])
+    phase_ratios  = TA(backend)(Float64, data["phase_ratios"])
+    particle_args = TA(backend).(Float64, data["particle_args"])
```
Comment (albert-de-montserrat) [inline docs/src/IO.md:38]:
```suggestion
data          = load("my_file.jld2")
particles     = TA(backend)(Float64, data["particles"])
phases        = TA(backend)(Float64, data["phases"])
phase_ratios  = TA(backend)(Float64, data["phase_ratios"])
particle_args = TA(backend).(Float64, data["particle_args"])
```

File: docs/src/mixed_CPU_GPU.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+This involves basically 4 steps:
+
+1) *At the top of the script*. The JustPIC backend must be set to CPU, in contrast with other employed packages (e.g. ParallelStencil):
+```julia
+const backend = JustPIC.CPUBackend 
+```
+
+2) *At memory allocation stage*. A copy of relevant CPU arrays must be allocated on the GPU memory. For example, phase ratios on mesh vertices:
+```julia
+phv_GPU = @zeros(Nc.x+1, Nc.y+1, Nc.z+1, celldims=(N_phases))
```
Comment (albert-de-montserrat) [inline docs/src/mixed_CPU_GPU.md:13]:
```suggestion
phv_GPU = @zeros(nx+1, ny+1, nz+1, celldims=(number_of_phases))
```

Perhaps the variable names are more obvious in this way.

File: docs/src/mixed_CPU_GPU.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+```
+where `N_phases` is the number of different material phases and `@zeros()` allocates on the GPU.
+
+Similarly, GPU arrays must be copied to CPU memory:
+```julia
+V_CPU = (
+        x      = zeros(Nc.x+1, Nc.y+2, Nc.z+2),
+        y      = zeros(Nc.x+2, Nc.y+1, Nc.z+2),
+        z      = zeros(Nc.x+2, Nc.y+2, Nc.z+1),
+    )
```
Comment (albert-de-montserrat) [inline docs/src/mixed_CPU_GPU.md:23]:
```suggestion
V_CPU = (
    x = zeros(nx+1, ny+2, nz+2),
    y = zeros(nx+2, ny+1, nz+2),
    z = zeros(nx+2, ny+2, nz+1),
)
```

File: docs/src/mixed_CPU_GPU.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+```
+where `zeros()` allocates on the CPU memory.
+
+4) *At each time step*. The particle will be stored on the CPU memory. It is hence necessary to transfer some information from the CPU to the GPU memory. For example, here's a transfer of phase proportions:
+
+```julia
+phv_GPU.data .= CuArray(phase_ratios.vertex).data
+```
+!!! we explicitly write `CuArray` - would be better to have something more explicit like `GPUArray` - is there such a thing?
```
Comment (albert-de-montserrat) [inline docs/src/mixed_CPU_GPU.md:32]: I guess the backend here depends either on what pkg one is using for the GPU backend abstraction, which does not match JPs backend now. With ParallelStencil this should be `Data.Array`, but this is different with KernelAbstractions. I would leave `CuArray`

**Review by albert-de-montserrat — COMMENTED** [review]
Thanks a lot!

File: docs/src/mixed_CPU_GPU.md (hunk 4 above)
Comment (tduretz) [reply docs/src/mixed_CPU_GPU.md:13]: perfect!

**Pull request merged**